File: src/com/eagle/programmar/Bash/Commands/Bash_Function.java
// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 15, 2022

package com.eagle.programmar.Bash.Commands;

import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnable;
import com.eagle.metrics.ArgumentsMetrics;
import com.eagle.metrics.CallMetrics;
import com.eagle.programmar.Bash.Bash_Element;
import com.eagle.programmar.Bash.Bash_EndOfLine;
import com.eagle.programmar.Bash.Bash_Syntax;
import com.eagle.programmar.Bash.Symbols.Bash_Function_Definition;
import com.eagle.programmar.Bash.Terminals.Bash_Keyword;
import com.eagle.scope.EagleScope;
import com.eagle.scope.EagleScope.EagleScopeInterface;
import com.eagle.tokens.AbstractFunction;
import com.eagle.tokens.TokenChooser;
import com.eagle.tokens.TokenList;
import com.eagle.tokens.TokenSequence;
import com.eagle.tokens.punctuation.PunctuationLeftBrace;
import com.eagle.tokens.punctuation.PunctuationLeftParen;
import com.eagle.tokens.punctuation.PunctuationRightBrace;
import com.eagle.tokens.punctuation.PunctuationRightParen;

public class Bash_Function extends TokenChooser
{
	public @CHOICE static class Bash_Function_Explicit extends TokenSequence
				implements AbstractFunction, EagleRunnable, EagleScopeInterface
	{
		public @S(10) @DOC("#index-functions_002c-shell") Bash_Keyword FUNCTION = new Bash_Keyword("function");
		public @S(20) Bash_Function_Definition id;
		public @S(30) @OPT Bash_FunctionParams params;
		public @S(40) @OPT Bash_EndOfLine eoln1;
		public @S(50) PunctuationLeftBrace leftBrace;
		public @S(60) Bash_EndOfLine eoln2;
		public @S(70) TokenList<Bash_Element> statements;
		public @S(80) PunctuationRightBrace rightBrace;

		public static class Bash_FunctionParams extends TokenSequence
		{
			public @S(10) PunctuationLeftParen leftParen;
			public @S(20) PunctuationRightParen rightParen;
		}
		
		public @SKIP CallMetrics _callMetrics = null;
		public @SKIP ArgumentsMetrics _argumentsMetrics = null;
		
		// Bash has a strange way of returning values
		public @SKIP int _exitStatus = 0;
		public @SKIP String _echoOutputs = null;

		private @SKIP EagleScope _scope = new EagleScope(this, Bash_Syntax.IS_CASE_SENSITIVE);

		@Override
		public EagleScope getScope()
		{
			return _scope;
		}
		
		@Override
		public void interpret(EagleInterpreter interpreter)
		{
			if (_callMetrics == null)
			{
				_callMetrics = new CallMetrics(interpreter._metrics, id.getValue(), id);
			}
			if (_argumentsMetrics == null)
			{
				_argumentsMetrics = new ArgumentsMetrics(interpreter._metrics, id.getValue(), id);
			}
		}
	}

	public @CHOICE static class Bash_Function_Implicit extends TokenSequence
	{
		public @S(10) Bash_Function_Definition fnName;
		public @S(20) PunctuationLeftParen leftParen;
		public @S(30) PunctuationRightParen rightParen;
		public @S(40) PunctuationLeftBrace leftBrace;
		public @S(50) Bash_EndOfLine eoln1;
		public @S(60) TokenList<Bash_Element> statements;
		public @S(70) PunctuationRightBrace rightBrace;
	}
}

File: csrc/com/eagle/programmar/Bash/Commands/Bash_Function.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 15, 2022

namespace com.eagle.programmar.Bash.Commands
{
	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnable = com.eagle.interpret.EagleRunnable;
	using ArgumentsMetrics = com.eagle.metrics.ArgumentsMetrics;
	using CallMetrics = com.eagle.metrics.CallMetrics;
	using Bash_Element = com.eagle.programmar.Bash.Bash_Element;
	using Bash_EndOfLine = com.eagle.programmar.Bash.Bash_EndOfLine;
	using Bash_Syntax = com.eagle.programmar.Bash.Bash_Syntax;
	using Bash_Function_Definition = com.eagle.programmar.Bash.Symbols.Bash_Function_Definition;
	using Bash_Keyword = com.eagle.programmar.Bash.Terminals.Bash_Keyword;
	using EagleScope = com.eagle.scope.EagleScope;
	using EagleScopeInterface = com.eagle.scope.EagleScope.EagleScopeInterface;
	using AbstractFunction = com.eagle.tokens.AbstractFunction;
	using TokenChooser = com.eagle.tokens.TokenChooser;
	using TokenList = com.eagle.tokens.TokenList;
	using TokenSequence = com.eagle.tokens.TokenSequence;
	using PunctuationLeftBrace = com.eagle.tokens.punctuation.PunctuationLeftBrace;
	using PunctuationLeftParen = com.eagle.tokens.punctuation.PunctuationLeftParen;
	using PunctuationRightBrace = com.eagle.tokens.punctuation.PunctuationRightBrace;
	using PunctuationRightParen = com.eagle.tokens.punctuation.PunctuationRightParen;

	public class Bash_Function : TokenChooser
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @CHOICE static class Bash_Function_Explicit extends com.eagle.tokens.TokenSequence implements com.eagle.tokens.AbstractFunction, com.eagle.interpret.EagleRunnable, com.eagle.scope.EagleScope.EagleScopeInterface
		public class Bash_Function_Explicit : TokenSequence, AbstractFunction, EagleRunnable, EagleScope.EagleScopeInterface
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) @DOC("#index-functions_002c-shell") com.eagle.programmar.Bash.Terminals.Bash_Keyword FUNCTION = new com.eagle.programmar.Bash.Terminals.Bash_Keyword("function");
			public @DOC("#index-functions_002c-shell") Bash_Keyword FUNCTION = new Bash_Keyword("function");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.programmar.Bash.Symbols.Bash_Function_Definition id;
			public Bash_Function_Definition id;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) @OPT Bash_FunctionParams params;
			public @OPT Bash_FunctionParams @params;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(40) @OPT Bash_EndOfLine eoln1;
			public @OPT Bash_EndOfLine eoln1;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(50) com.eagle.tokens.punctuation.PunctuationLeftBrace leftBrace;
			public PunctuationLeftBrace leftBrace;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(60) com.eagle.programmar.Bash.Bash_EndOfLine eoln2;
			public Bash_EndOfLine eoln2;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(70) com.eagle.tokens.TokenList<com.eagle.programmar.Bash.Bash_Element> statements;
			public TokenList<Bash_Element> statements;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(80) com.eagle.tokens.punctuation.PunctuationRightBrace rightBrace;
			public PunctuationRightBrace rightBrace;

			public static class Bash_FunctionParams extends TokenSequence
			{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.tokens.punctuation.PunctuationLeftParen leftParen;
				public PunctuationLeftParen leftParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.punctuation.PunctuationRightParen rightParen;
				public PunctuationRightParen rightParen;
			}

// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP CallMetrics _callMetrics = null;
			public CallMetrics _callMetrics = null;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP ArgumentsMetrics _argumentsMetrics = null;
			public ArgumentsMetrics _argumentsMetrics = null;

			// Bash has a strange way of returning values
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP int _exitStatus = 0;
			public int _exitStatus = 0;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP String _echoOutputs = null;
			public string _echoOutputs = null;

// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: private @SKIP EagleScope _scope = new com.eagle.scope.EagleScope(this, com.eagle.programmar.Bash.Bash_Syntax.IS_CASE_SENSITIVE);
			private EagleScope _scope = new EagleScope(this, Bash_Syntax.IS_CASE_SENSITIVE);

			public EagleScope Scope
			{
				return _scope;
			}

			public void interpret(EagleInterpreter interpreter)
			{
				if (_callMetrics == null)
				{
					_callMetrics = new CallMetrics(interpreter._metrics, id.getValue(), id);
				}
				if (_argumentsMetrics == null)
				{
					_argumentsMetrics = new ArgumentsMetrics(interpreter._metrics, id.getValue(), id);
				}
			}
		}

// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @CHOICE static class Bash_Function_Implicit extends com.eagle.tokens.TokenSequence
		public static class Bash_Function_Implicit extends TokenSequence
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Bash.Symbols.Bash_Function_Definition fnName;
			public Bash_Function_Definition fnName;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.punctuation.PunctuationLeftParen leftParen;
			public PunctuationLeftParen leftParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) com.eagle.tokens.punctuation.PunctuationRightParen rightParen;
			public PunctuationRightParen rightParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(40) com.eagle.tokens.punctuation.PunctuationLeftBrace leftBrace;
			public PunctuationLeftBrace leftBrace;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(50) com.eagle.programmar.Bash.Bash_EndOfLine eoln1;
			public Bash_EndOfLine eoln1;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(60) com.eagle.tokens.TokenList<com.eagle.programmar.Bash.Bash_Element> statements;
			public TokenList<Bash_Element> statements;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(70) com.eagle.tokens.punctuation.PunctuationRightBrace rightBrace;
			public PunctuationRightBrace rightBrace;
		}
	}

}
